File: truezip-kernel/src/main/java/de/schlichtherle/truezip/socket/ResourceAccountant.java
/*
 * Copyright (C) 2006-2011 Schlichtherle IT Services
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.schlichtherle.truezip.socket;

import de.schlichtherle.truezip.util.ExceptionHandler;
import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
import edu.umd.cs.findbugs.annotations.NonNull;
import edu.umd.cs.findbugs.annotations.Nullable;
import java.io.Closeable;
import java.io.IOException;
import java.lang.ref.ReferenceQueue;
import java.lang.ref.WeakReference;
import java.util.Iterator;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.logging.Level;
import java.util.logging.Logger;
import net.jcip.annotations.ThreadSafe;

/**
 * Accounts for {@link Closeable} resources used in multiple threads while
 * holding a synchronization lock provided to the constructor.
 * <p>
 * In order to start accounting for a closeable resource,
 * call {@link #startAccountingFor(Closeable)}.
 * In order to stop accounting for a closeable resource,
 * call {@link #stopAccountingFor(Closeable)}.
 *
 * @param   <R> The type of the resources.
 * @author  Christian Schlichtherle
 * @version $Id$
 */
@ThreadSafe
@DefaultAnnotation(NonNull.class)
final class ResourceAccountant implements Closeable {

    private static final String CLASS_NAME
            = ResourceAccountant.class.getName();
    private static final Logger logger
            = Logger.getLogger(CLASS_NAME, CLASS_NAME);

    private final Condition condition;
    private final Lock lock;

    /**
     * The pool of all accounted closeable resources.
     * The weak hash map allows the garbage collector to pick up a closeable
     * resource if there are no more references to it.
     */
    private volatile @Nullable Map<Closeable, ResourceReference>
            threads = new WeakHashMap<Closeable, ResourceReference>();

    /**
     * Constructs a new resource accountant with the given synchronization
     * lock.
     * 
     * @param lock the synchronization lock to use for accounting resources.
     *             Though not required by the use in this class, this
     *             parameter should normally be an instance of
     *             {@link ReentrantLock} because chances are that it gets
     *             locked recursively.
     */
    public ResourceAccountant(final Lock lock) {
        this.condition = lock.newCondition();
        this.lock = lock;
    }

    /**
     * Returns {@code true} if and only if this resource accountant has
     * been {@link #close() closed}.
     * 
     * @return {@code true} if and only if this resource accountant has
     *         been {@link #close() closed}.
     */
    public boolean isClosed() {
        return null == this.threads;
    }

    /**
     * Starts accounting for the given closeable resource.
     * 
     * @param  resource the closeable resource to start accounting for.
     * @return {@code true} if and only if the given closeable resource is not
     *         already accounted for.
     * @throws IllegalStateException if this resource accountant has already
     *         been closed.
     */
    public boolean startAccountingFor(final Closeable resource) {
        this.lock.lock();
        try {
            if (isClosed())
                throw new IllegalStateException("Already closed!");
            if (this.threads.containsKey(resource))
                return false;
            this.threads.put(resource, new ResourceReference(resource));
            return true;
        } finally {
            this.lock.unlock();
        }
    }

    /**
     * Stops accounting for the given closeable resource.
     * This method should be called from the implementation of
     * {@link Closeable#close()} in the given closeable resource.
     * 
     * @param  resource the closeable resource to stop accounting for.
     * @return {@code true} if and only if the given closeable resource was
     *         accounted for.
     */
    public boolean stopAccountingFor(final Closeable resource) {
        this.lock.lock();
        try {
            if (isClosed())
                return false;
            final ResourceReference ref = this.threads.remove(resource);
            if (null == ref)
                return false;
            if (null != ref.get() && ref.getOwner() != Thread.currentThread())
                this.condition.signal();
            return true;
        } finally {
            this.lock.unlock();
        }
    }

    /**
     * Waits until all closeable resources which have been started accounting
     * for in <em>other</em> threads get stopped accounting for or a timeout
     * occurs.
     * If the current thread is interrupted while waiting,
     * then immediately a warn message is logged using
     * {@code java.util.logging} and control is returned to the caller.
     * <p>
     * Upon return of this method, threads may immediately start accounting
     * for closeable resources again unless the caller also synchronizes on the
     * lock provided to the constructor - use with care!
     *
     * @param  the number of milliseconds to await the closing of resources
     *         which have been accounted for in <em>other</em> threads.
     *         If this is {@code 0}, then there is no timeout for waiting.
     * @return The number of <em>all</em> accounted closeable resources.
     */
    public int waitStop(final long timeout) {
        this.lock.lock();
        try {
            if (isClosed())
                return 0;
            final long start = System.currentTimeMillis();
            try {
                while (this.threads.size() > threadLocalResources()) {
                    long toWait;
                    if (timeout > 0) {
                        toWait = timeout - (System.currentTimeMillis() - start);
                        if (toWait <= 0)
                            break;
                    } else {
                        toWait = 0;
                    }
                    if (!this.condition.await(timeout, TimeUnit.MILLISECONDS))
                        break;
                }
            } catch (InterruptedException ex) {
                logger.log(Level.WARNING, "interrupted", ex);
            }
            return this.threads.size();
        } finally {
            this.lock.unlock();
        }
    }

    /**
     * Returns the number of closeable resources which have been accounted for
     * in the <em>current</em> thread.
     * This method must be externally synchronized!
     */
    private int threadLocalResources() {
        assert !isClosed();
        int n = 0;
        final Thread currentThread = Thread.currentThread();
        for (final ResourceReference ref : this.threads.values())
            if (ref.getOwner() == currentThread)
                n++;
        return n;
    }

    /**
     * For each accounted closeable resource,
     * stops accounting for it and closes it.
     * <p>
     * Upon return of this method, threads may immediately start accounting
     * for closeable resources again unless the caller also synchronizes on the
     * lock provided to the constructor - use with care!
     */
    public <X extends Exception>
    void closeAll(final ExceptionHandler<? super IOException, X> handler)
    throws X {
        this.lock.lock();
        try {
            if (isClosed())
                return;
            for (final Iterator<Closeable> i = this.threads.keySet().iterator(); i.hasNext(); ) {
                try {
                    final Closeable closeable = i.next();
                    i.remove();
                    // This may trigger another removal, but it should cause no
                    // ConcurrentModificationException because the closeable is no
                    // more present in the map.
                    closeable.close();
                } catch (IOException ex) {
                    handler.warn(ex);
                }
            }
            assert this.threads.isEmpty();
        } finally {
            this.lock.unlock();
        }
    }

    /**
     * Closes this resource accountant.
     *
     * @see    #waitStop(long)
     * @see    #closeAll
     * @throws IOException If any accounted closeable resources are detected.
     */
    @Override
    public void close() throws IOException {
        this.lock.lock();
        try {
            if (isClosed())
                return;
            final int size = this.threads.size();
            if (0 != size)
                throw new IOException("There are still " + size + " accounted closeable resources!");
            this.threads = null;
        } finally {
            this.lock.unlock();
        }
    }

    /**
     * Accounts for a resource and its owner, which is the thread in which the
     * account is created.
     */
    private final class ResourceReference extends WeakReference<Closeable> {
        private final Thread owner = Thread.currentThread();

        ResourceReference(Closeable resource) {
            super(resource, ResourceCollector.queue);
        }

        Thread getOwner() {
            return owner;
        }

        void notifyAccountant() {
            final Lock lock = ResourceAccountant.this.lock;
            lock.lock();
            try {
                ResourceAccountant.this.condition.signal();
            } finally {
                lock.unlock();
            }
        }
    }

    private static final class ResourceCollector extends Thread {
        static {
            new ResourceCollector().start();
        }

        static final ReferenceQueue<Closeable>
                queue = new ReferenceQueue<Closeable>();

        private ResourceCollector() {
            super("TrueZIP Resource Collector");
            setDaemon(true);
        }

        @Override
        public void run() {
            while (true) {
                try {
                    ((ResourceReference) queue.remove()).notifyAccountant();
                } catch (InterruptedException ex) {
                    logger.log(Level.WARNING, "interrupted", ex);
                }
            }
        }
    }
}

File: truevfs-kernel/src/main/scala/de/schlichtherle/truevfs/kernel/ResourceAccountant.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel

import collection.JavaConverters._
import java.io._
import java.util.concurrent._
import java.util.concurrent.locks._
import javax.annotation._
import javax.annotation.concurrent._
import net.truevfs.kernel.util._
import scala.util.control._

/** Accounts for [[java.io.Closeable]] resources.
  * 
  * For synchronization, each accountant uses a lock which has to be provided
  * to its constructor.
  * In order to start accounting for a closeable resource, call `start`.
  * In order to stop  accounting for a closeable resource, call `stop`.
  * 
  * Note that you ''must make sure'' not to use two instances of this class
  * which share the same lock!
  * Otherwise `waitOtherThreads` will not work as designed!
  * 
  * @param  lock the lock to use for accounting resources.
  * @see    FsResourceController
  * @author Christian Schlichtherle
  */
@ThreadSafe
private final class ResourceAccountant(lock: Lock) {
  import ResourceAccountant._

  private[this] val condition = lock.newCondition

  /** Starts accounting for the given closeable resource.
    * 
    * @param resource the closeable resource to start accounting for.
    */
  def startAccountingFor(@WillCloseWhenClosed resource: Closeable) {
    accounts += resource -> Account(this)
  }

  /** Stops accounting for the given closeable resource.
    * This method should be called from the implementation of the `close`
    * method of the given [[java.io.Closeable]].
    * 
    * @param resource the closeable resource to stop accounting for.
    */
  def stopAccountingFor(@WillNotClose resource: Closeable) {
    accounts.remove(resource) foreach { _ =>
      lock lock()
      try {
        condition signalAll()
      } finally {
        lock unlock()
      }
    }
  }

  /** Waits until all closeable resources which have been started accounting
    * for by ''other'' threads get stopped accounting for or a timeout occurs
    * or the current thread gets interrupted, whatever happens first.
    * 
    * Waiting for such resources can get cancelled immediately by interrupting
    * the current thread.
    * Unless the number of closeable resources which have been accounted for
    * by ''all'' threads is zero, this will leave the interrupt status of the
    * current thread cleared.
    * If no such foreign resources exist, then interrupting the current thread
    * does not have any effect.
    * 
    * Upon return of this method, threads may immediately start accounting
    * for closeable resources again unless the caller has acquired the lock
    * provided to the constructor - use with care!
    * 
    * Note that this method '''will not work''' if any two instances of this
    * class share the same lock provided to their constructor!
    *
    * @param timeout the number of milliseconds to await the closing of
    *        resources which have been accounted for by ''other''
    *        threads once the lock has been acquired.
    *        If this is non-positive, then there is no timeout for waiting.
    */
  def waitOtherThreads(timeout: Long) {
    lock lock()
    try {
      try {
        var toWait = TimeUnit.MILLISECONDS toNanos timeout
        val mybreaks = new Breaks
        import mybreaks.{break, breakable}
        breakable {
          while (resources.isBusy) {
            if (0 < timeout) {
              if (0 >= toWait) break
              toWait = condition awaitNanos toWait
            } else {
              condition await()
            }
          }
        }
      } catch {
        case _: InterruptedException =>
          // Fix rare racing condition between Thread.interrupt() and
          // Condition.signalAll() events.
          if (0 == resources.total)
            Thread.currentThread.interrupt()
      }
    } finally {
      lock unlock()
    }
  }

  /** Returns the number of closeable resources which have been accounted for.
    * The first element contains the number of closeable resources which have
    * been created by the current thread (''local'').
    * The second element contains the number of closeable resources which have
    * been created by all threads (''total'').
    * Mind that this value may reduce concurrently, even while the lock is
    * held, so it should ''not'' get cached!
    * 
    * @return The number of closeable resources which have been accounted for.
    */
  def resources = {
    val currentThread = Thread.currentThread
    var local, total = 0
    for (account <- accounts.values if account.accountant eq this) {
      if (account.owner eq currentThread) local += 1
      total += 1
    }
    Resources(local, total)
  }

  /** For each accounted closeable resource, stops accounting for it and closes
    * it.
    * Upon return of this method, threads may immediately start accounting for
    * closeable resources again unless the caller also locks the lock provided
    * to the constructor - use with care!
    */
  def closeAllResources() {
    lock lock()
    try {
      val builder = new SuppressedExceptionBuilder[IOException]
      for ((closeable, account) <- accounts if account.accountant eq this) {
        accounts -= closeable
        try {
          // This should trigger an attempt to remove the closeable from the
          // map, but it can cause no ConcurrentModificationException because
          // the entry is already removed and a ConcurrentHashMap doesn't do
          // that anyway.
          // Note that this method may throw an IOException or a
          // RuntimeException, e.g. a NeedsLockRetryException!
          closeable close()
        } catch {
          case ex: ControlFlowException =>
            assert(ex.isInstanceOf[NeedsLockRetryException], ex)
            throw ex
          case ex: IOException =>
            builder warn ex // could throw an IOException!
        }
      }
      builder check()
    } finally {
      condition signalAll()
      lock unlock()
    }
  }
}

private object ResourceAccountant {

  /** The map of all accounted closeable resources.
    * The initial capacity for the hash map accounts for the number of
    * available processors, a 90% blocking factor for typical I/O and a 2/3
    * map resize threshold.
    */
  private val accounts = {
    val initialCapacity =
      HashMaps.initialCapacity(Runtime.getRuntime.availableProcessors() * 10);
    new ConcurrentHashMap[Closeable, Account](
      initialCapacity, 0.75f, initialCapacity).asScala
  }

  private final case class Account(accountant: ResourceAccountant) {
    val owner = Thread.currentThread
  } // Account

  final case class Resources(local: Int, total: Int) {
    def isBusy = local < total
  } // Resources
}
